### ISSUE TITLE
[docker][agentbeat] - insufficient capabilities for `agentbeat` causes agent degradation

### ISSUE THREAD

**Issue by VihasMakwana** [elastic/elastic-agent#5269]
### Current behaviour:
 https://github.com/elastic/elastic-agent/blob/a1fd2c9d9a7cc3931f8f1d452a730b6d216f8bd3/dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl#L60
 assigns some capabilities in `permitted` set. 

While running `unprivileged` mode,  agentbeat needs to have `cap_sys_ptrace` and `cap_dac_override` to access the process related metrics and summary. Due to above `setcap`, current effective capabilities are empty which causes the agent degradation (verifiable via `/proc/[agentbeat pid]/status`.

There are two ways which can fix this bug:
- Add `cap_sys_ptrace` and `cap_dac_override` to effective set, [example commit](https://github.com/elastic/elastic-agent/compare/main...VihasMakwana:fix-capabilities-agentbeat?expand=1#diff-21d89ed5947be16887bfdba934bd3af27ef7252551adccd996a9bdc961222f39)
    - Tested this out and works as expected.
- Update `agentbeat` to elevate the effective capabilities to match the bounding set.
    - We need to do this before `seccomp`, as it sets the `no_new_privs` flag.
   

I'd appreciate your thoughts on this.

**Comment by pierrehilbert** [comment]
@blakerouse could you please have a look here?

**Comment by blakerouse** [comment]
> 
> * Add `cap_sys_ptrace` and `cap_dac_override` to effective set, [example commit](https://github.com/elastic/elastic-agent/compare/main...VihasMakwana:fix-capabilities-agentbeat?expand=1#diff-21d89ed5947be16887bfdba934bd3af27ef7252551adccd996a9bdc961222f39)

This seems like the simplest solution without adding the logic to agentbeat to elevate, but what are the downsides? Does this mean that the binary will fail to run in the bounding set doesn't have these capabilities? If it means that it will fail to run if the bounding set is restricted, then this is not an option. If it still runs just limited by the defined bounding set, then this is the simplest path.

>   
>   * Tested this out and works as expected.
> * Update `agentbeat` to elevate the effective capabilities to match the bounding set.
>   
>   * We need to do this before `seccomp`, as it sets the `no_new_privs` flag.

We should do this if the limitation above occurs.

**Comment by VihasMakwana** [comment]
> the binary will fail to run in the bounding set doesn't have these capabilities? If it means that it will fail to run if the bounding set is restricted, then this is not an option. If it still runs just limited by the defined bounding set, then this is the simplest path.

If `cap_sys_ptrace` and `cap_dac_override` are not available, the binary will still execute but with reduced functionality. The `system.process` integration will be marked as `DEGRADED`, other components will be healthy.

**Comment by blakerouse** [comment]
> If `cap_sys_ptrace` and `cap_dac_override` are not available, the binary will still execute but with reduced functionality. The `system.process` integration will be marked as `DEGRADED`, other components will be healthy.

Then set it like you have in the example commit, simplest solution.

**Comment by VihasMakwana** [comment]
@blakerouse Sorry, in my previous comment I meant about the current behaviour (without my commit). 

But if my commit is applied locally and I build a local docker image, I **need** to add these capabilities while running on kubernetes standalone mode, or else it throws "operation not permitted" and binary will not start.

I think elevating is the safest way here. Let me know what do you think.

I think the overall solution would looks something like following:
- Add `cap_sys_ptrace` and `cap_dac_override` only in `permitted` set.
- Update `agentbeat` to elevate to bounding set if running as non-root.

**Comment by VihasMakwana** [comment]
@blakerouse agree with your point to move it to a common library.
But I believe we should merge https://github.com/elastic/elastic-agent/pull/5271 and https://github.com/elastic/beats/pull/40466 for now to unblock the CI.

I'll create a separate issue to move the common code a https://github.com/elastic/elastic-agent-libs

**Comment by VihasMakwana** [comment]
Let me know what do you think.

**Comment by VihasMakwana** [comment]
Closing this. See https://github.com/elastic/beats/pull/40466#issuecomment-2286307998
